zipline/assets/assets.py

```diff
@@ -0,0 +1,540 @@
+#
+# Copyright 2015 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABCMeta
+from itertools import chain
+from numbers import Integral
+import numpy as np
+import operator
+
+from logbook import Logger
+import pandas as pd
+from pandas.tseries.tools import normalize_date
+from six import with_metaclass
+
+from zipline.errors import (
+    SymbolNotFound,
+    MultipleSymbolsFound,
+    SidNotFound,
+    ConsumeAssetMetaDataError,
+    InvalidAssetType
+)
+from zipline.utils import tradingcalendar
+from zipline.assets._assets import (
+    Asset, Equity, Future
+)
+
+log = Logger('assets.py')
+
+
+class AssetFinder(object):
+
+    shared_caches = {'by_sid': {}, 'by_symbol': {}, 'fuzzy_match': {}}
+
+    def __init__(self,
+                 metadata,
+                 trading_calendar=tradingcalendar,
+                 force_populate=False):
+
+        # Any particular instance of AssetFinder should be
+        # consistent throughout its lifetime, so we grab a reference
+        # to our cache now. That way, if the cache is refreshed later,
+        # our instance will continue to use the old one.
+        self.cache = self.shared_caches['by_sid']
```

Comment:
This may not be worth changing now, but I don't think we should be caching these values on the class.  If we're only ever going to instantiate one of these per simulation, I'd rather just treat this as a singleton the way we do with TradingEnvironment.